### PR TITLE
feat(libpng): add parsing variable to libpng

### PR DESCRIPTION
**Pull request by lhdjply**
### Description of the feature or fix

feat(libpng): add parsing variable to libpng

### Notes
- Update the [Documentation](https://github.com/lvgl/lvgl/tree/master/docs) if needed.
- Add [Examples](https://github.com/lvgl/lvgl/tree/master/examples) if relevant.
- Add [Tests](https://github.com/lvgl/lvgl/blob/master/tests/README.md) if applicable.
- If you added new options to `lv_conf_template.h` run [lv_conf_internal_gen.py](https://github.com/lvgl/lvgl/blob/master/scripts/lv_conf_internal_gen.py) and update [Kconfig](https://github.com/lvgl/lvgl/blob/master/Kconfig).
- Run `scripts/code-format.py` ([astyle](http://astyle.sourceforge.net/install.html) version [v3.4.12](https://github.com/szepeviktor/astyle/releases/tag/v3.4.12) needs to be installed) and follow the [Code Conventions](https://docs.lvgl.io/master/CODING_STYLE.html).
- Mark the Pull request as [Draft](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-stage-of-a-pull-request) while you are working on the first version, and mark is as _Ready_ when it's ready for review.
- When changes were requested, [re-request review](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review) to notify the maintainers.
- Help us to review this Pull Request! Anyone can [approve or request changes](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/approving-a-pull-request-with-required-reviews).

### CODE REVIEW

File: src/libs/libpng/lv_libpng.c [hunk 1]
```diff
@@ -109,6 +110,33 @@ static lv_result_t decoder_info(lv_image_decoder_t * decoder, lv_image_decoder_d
 
         return LV_RESULT_OK;
     }
+    /*If it's a PNG file in a  C array...*/
+    else if(src_type == LV_IMAGE_SRC_VARIABLE) {
+        const lv_image_dsc_t * img_dsc = src;
+        const uint32_t data_size = img_dsc->data_size;
+        const uint32_t * size = ((uint32_t *)img_dsc->data) + 4;
```
Comment (XuNeo) [inline src/libs/libpng/lv_libpng.c]: Let's extract the common code and reuse it for both variable and file.

File: src/libs/libpng/lv_libpng.c [hunk 2]
```diff
@@ -236,27 +259,42 @@ static uint8_t * alloc_file(const char * filename, uint32_t * size)
...
+    if(dsc->src_type == LV_IMAGE_SRC_FILE) {
+        if(lv_strcmp(lv_fs_get_ext(dsc->src), "png") == 0) {              /*Check the extension*/
+            png_data = alloc_file(dsc->src, &png_data_size);
+            if(png_data == NULL) {
+                LV_LOG_WARN("can't load file: %s", (const char *)dsc->src);
+                return NULL;
+            }
+        }
+        else
+            return NULL;
```
Comment (XuNeo) [inline src/libs/libpng/lv_libpng.c]: Return early on error and keep function body focuses on the main job.

```suggestion
        if(lv_strcmp(lv_fs_get_ext(dsc->src), "png") != 0) {              /*Check the extension*/
            return NULL;
        }

        png_data = alloc_file(dsc->src, &png_data_size);
        if(png_data == NULL) {
            LV_LOG_WARN("can't load file: %s", (const char *)dsc->src);
            return NULL;
        }
```

File: src/libs/libpng/lv_libpng.c (hunk 1 above)
Comment (lhdjply) [reply src/libs/libpng/lv_libpng.c:117]: This code was copied from lv_lodepng, so should lv_lodepng be modified in the same way as well?

**Comment by lhdjply** [comment]
@kisvegabor Can we enable non-Collaborators to request reviews from Collaborators? Sometimes, when PRs go unreviewed, the only option is to @ mention them, which can be inconvenient.

image

File: src/libs/libpng/lv_libpng.c (hunk 1 above)
Comment (lhdjply) [reply src/libs/libpng/lv_libpng.c]: Done

**Comment by kisvegabor** [comment]
> Can we enable non-Collaborators to request reviews from Collaborators? Sometimes, when PRs go unreviewed, the only option is to @ mention them, which can be inconvenient.

I haven't found such an option. Is it really configurable?

**Review by kisvegabor — DISMISSED** [review]

**Comment by lhdjply** [comment]
> I haven't found such an option. Is it really configurable?

I see that other repositories can achieve this, for example, thorvg.

https://github.com/thorvg/thorvg/pull/2316

image

File: src/libs/libpng/lv_libpng.c [hunk 3]
```diff
@@ -83,7 +83,10 @@ static lv_result_t decoder_info(lv_image_decoder_t * decoder, lv_image_decoder_d
 {
     LV_UNUSED(decoder); /*Unused*/
 
+    const void * src = dsc->src;
     lv_image_src_t src_type = dsc->src_type;          /*Get the source type*/
+    uint32_t * size;
+    const uint8_t magic[] = {0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a};
```
Comment (XuNeo) [inline src/libs/libpng/lv_libpng.c]:
```suggestion
    static const uint8_t magic[] = {0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a};
```

**Comment by kisvegabor** [comment]
Maybe @hermet or @XuNeo can tell more about it 🙂 @XuNeo are you part of the ThorVG in some form?

**Comment by hermet** [comment]
@kisvegabor maybe?
image

**Comment by kisvegabor** [comment]
Thank you @hermet, but both are enabled. :slightly_frowning_face:

**Comment by XuNeo** [comment]
No worry though, after PR has been reviewed, owner can re-request review after new changes.

File: src/libs/libpng/lv_libpng.c [hunk 4]
```diff
@@ -109,6 +111,33 @@ static lv_result_t decoder_info(lv_image_decoder_t * decoder, lv_image_decoder_d
...
+        else {
+            header->w = (int32_t)((size[0] & 0xff000000) >> 24) + ((size[0] & 0x00ff0000) >> 8);
+        }
+
+        if(img_dsc->header.h) {
+            header->h = img_dsc->header.h;         /*Save the color height*/
+        }
+        else {
+            header->h = (int32_t)((size[1] & 0xff000000) >> 24) + ((size[1] & 0x00ff0000) >> 8);
+        }
```
Comment (XuNeo) [inline src/libs/libpng/lv_libpng.c]: Can we re-use the same logic for file type of image?

File: src/libs/libpng/lv_libpng.c [hunk 5]
```diff
@@ -274,8 +313,9 @@ static lv_draw_buf_t * decode_png_file(lv_image_decoder_dsc_t * dsc, const char
     lv_draw_buf_t * decoded;
     decoded = lv_draw_buf_create_user(image_cache_draw_buf_handlers, image.width, image.height, cf, LV_STRIDE_AUTO);
     if(decoded == NULL) {
-        LV_LOG_ERROR("alloc PNG_IMAGE_SIZE(%" LV_PRIu32 ") failed: %s", (uint32_t)PNG_IMAGE_SIZE(image), filename);
-        lv_free(data);
+        LV_LOG_ERROR("alloc PNG_IMAGE_SIZE(%" LV_PRIu32 ")", (uint32_t)PNG_IMAGE_SIZE(image));
```
Comment (XuNeo) [inline src/libs/libpng/lv_libpng.c]: Keep the log of filename so it's easier to see what's wrong.

File: src/libs/libpng/lv_libpng.c (hunk 4 above)
Comment (lhdjply) [reply src/libs/libpng/lv_libpng.c]: I merged the file and VARIABLE. I applied the same modification method to lodepng.

File: examples/libs/libpng/index.rst [hunk 6]
```diff
@@ -1,4 +1,4 @@
-Open a PNG image from file
+Open a PNG image from file and variable
```
Comment (kisvegabor) [inline examples/libs/libpng/index.rst:1]: The decoration under the title should have the same amount of characters as the title.
```rst
Open a PNG image from file and variable
---------------------------------------
```

**Review by kisvegabor — COMMENTED** [review]
Okay for me, just one comment.

File: examples/libs/libpng/index.rst (hunk 6 above)
Comment (lhdjply) [reply examples/libs/libpng/index.rst:1]: Done

**Review by kisvegabor — APPROVED** [review]

**Review by XuNeo — APPROVED** [review]

**Pull request merged**